### ISSUE TITLE
jasper-2.0.14 heap-buffer-overflow (out-of-bound-read) 

### ISSUE THREAD

**Issue by littleroach110** [jasper-software/jasper#164]
# Overview

I have found a vulnerability in jasper latest version 2.0.14. It is a heap-buffer-overflow vulnerability and can cause out-of-bound read.

# Analysis and PoC

The attachment is zip file which includes my detail analysis report and PoC file. In order to avoid disclosing it before patch is released, I have encrypted it and developers can communicate with me to get the password.

### Decypted

[jasper-heap-buffer-overflow_decypted.zip](https://github.com/mdadams/jasper/files/1524519/jasper-heap-buffer-overflow_decypted.zip)

### No Decypted ZIP File
[jasper-heap-buffer-overflow.zip](https://github.com/mdadams/jasper/files/1541902/jasper-heap-buffer-overflow.zip)

# Author

Name： Qiang Li @VARAS of IIE
Email：littleroach110@gmail.com
Org: IIE (http://iie.ac.cn)

**Comment by thoger** [comment]
This is related to 58ba0365 / issue #120 that was fixed in 2.0.13.  The check added ensures there's at least one component in the image to encode.  However, for certain color space families the code assumes at least 3 components were provided:

https://github.com/mdadams/jasper/blob/version-2.0.14/src/libjasper/jp2/jp2_enc.c#L295-L312

I assume that code should be extended to error-out if fewer than 3 components are available in the image.

**Comment by mdadams** [comment]
The above zip file from the original bug reporter appears to be password protected.  It cannot do much with a password protected file. :(

**Comment by littleroach110** [comment]
I update the issue and upload the no decypted zip file. I hope it can provide some help to owner and contributor.

**Comment by twelveand0** [comment]
Hi @mdadams, may i know the processing progress of this issue. Thanks

**Comment by jubalh** [comment]
@littleroach110 did this one get a CVE assigned?

**Comment by jubalh** [comment]
In Jasper 2.0.19 I get:

```
imginfo -f ~/jasper/poc-hea-buffer
jp2 1 32818 50 8 1640900
```

**Comment by MaxKellermann** [comment]
This bug was fixed by 03db7c81f6a8a92d896249bc673877749987fd7a (CVE-2018-20570).
